File: Sprint6/E/C++/main.cpp
// https://contest.yandex.ru/contest/25069/problems/E/
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void dfs(int v, vector<vector<int>>& g, vector<bool>& visited, vector<int>& component) {
    visited[v] = true;
    component.push_back(v);
    for (const auto& u : g[v]) {
        if (!visited[u]) {
            dfs(u, g, visited, component);
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<vector<int>> g(n);
    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        g[u - 1].push_back(v - 1);
        g[v - 1].push_back(u - 1);
    }

    vector<bool> visited(n, false);
    vector<vector<int>> components;

    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            vector<int> component;
            dfs(i, g, visited, component);
            sort(component.begin(), component.end());
            components.push_back(component);
        }
    }

    cout << components.size() << endl;
    for (const auto& component : components) {
        for (const auto& v : component) {
            cout << v + 1 << ' ';
        }
        cout << endl;
    }

    return 0;
}

File: Sprint6/E/Rust/src/main.rs
// https://contest.yandex.ru/contest/25069/problems/E/

use std::io::{self, BufRead};

fn dfs(v: usize, g: &Vec<Vec<usize>>, visited: &mut Vec<bool>, component: &mut Vec<usize>) {
    visited[v] = true;
    component.push(v);
    for &u in &g[v] {
        if !visited[u] {
            dfs(u, g, visited, component);
        }
    }
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let first_line = lines.next().unwrap().unwrap();
    let nm: Vec<usize> = first_line
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();
    let (n, m) = (nm[0], nm[1]);

    let mut g: Vec<Vec<usize>> = vec![vec![]; n];

    for _ in 0..m {
        let edge: Vec<usize> = lines
            .next()
            .unwrap()
            .unwrap()
            .split_whitespace()
            .map(|x| x.parse::<usize>().unwrap() - 1)
            .collect();

        let (u, v) = (edge[0], edge[1]);
        g[u].push(v);
        g[v].push(u);
    }

    let mut visited: Vec<bool> = vec![false; n];
    let mut components: Vec<Vec<usize>> = Vec::new();

    for i in 0..n {
        if !visited[i] {
            let mut component: Vec<usize> = Vec::new();
            dfs(i, &g, &mut visited, &mut component);
            component.sort();
            components.push(component);
        }
    }

    println!("{}", components.len());
    for component in &components {
        for &v in component {
            print!("{} ", v + 1);
        }
        println!();
    }
}
